libevm/VM.cpp

```diff
@@ -129,35 +123,46 @@ void VM::logGasMem()
 
 void VM::fetchInstruction()
 {
-		m_op = Instruction(m_pc < m_codeSpace.size() ? m_code[m_pc] : 0);
-		const InstructionMetric& metric = c_metrics[static_cast<size_t>(m_op)];
-		checkStack(metric.args, metric.ret);
-
-		// FEES...
-		m_runGas = toUint64(m_schedule->tierStepGas[metric.gasPriceTier]);
-		m_newMemSize = m_mem.size();
-		m_copyMemSize = 0;
+	m_op = Instruction(m_code[m_pc]);
+	const InstructionMetric& metric = c_metrics[static_cast<size_t>(m_op)];
+	checkStack(metric.args, metric.ret);
+
+	// FEES...
+	m_runGas = toUint64(m_schedule->tierStepGas[metric.gasPriceTier]);
+	m_newMemSize = m_mem.size();
+	m_copyMemSize = 0;
 }
 
 
 ///////////////////////////////////////////////////////////////////////////////
 //
 // interpreter entry point
 
-bytesConstRef VM::execImpl(u256& io_gas, ExtVMFace& _ext, OnOpFunc const& _onOp)
+bytesConstRef VM::execImpl(u256& _io_gas, ExtVMFace& _ext, OnOpFunc const& _onOp)
 {
-	m_io_gas = &io_gas;
+	io_gas = &_io_gas;
+	m_io_gas = uint64_t(_io_gas);
 	m_ext = &_ext;
 	m_schedule = &m_ext->evmSchedule();
-	m_onOp = &_onOp;
-	m_onFail = &VM::doOnOperation;
-
-	// trampoline to minimize depth of call stack when calling out
-	m_bounce = &VM::initEntry;
-	do
-		(this->*m_bounce)();
-	while (m_bounce);
+	m_onOp = _onOp;
+	m_onFail = &VM::onOperation;
+	
+	try
+	{
+		// trampoline to minimize depth of call stack when calling out
+		m_bounce = &VM::initEntry;
+		do
+			(this->*m_bounce)();
+		while (m_bounce);
+		
+	}
+	catch (...)
+	{
+		*io_gas = m_io_gas;
```

Comment:
Everything caught here is rethrown.  The purpose is just to set the 256-bit _io_gas reference here, so that I don't have to set it via 256-bit arithmetic everywhere else - if I didn't catch everything here _io_gas could be wrong on some exceptions.  So the behavior should look just the same to the caller as is did before I introduced the try/catch.